metricbeat/module/windows/perfmon/reader.go

```diff
@@ -94,8 +93,38 @@ func NewReader(config Config) (*Reader, error) {
 	return r, nil
 }
 
+// RefreshCounterPaths will recheck for any new instances and add them to the counter list
+func (r *Reader) RefreshCounterPaths() error {
+	for _, counter := range r.config.CounterConfig {
+		if !strings.Contains(counter.Query, "*") {
+			continue
+		}
+		childQueries, err := r.query.ExpandWildCardPath(counter.Query)
+		if err == nil && len(childQueries) >= 1 && !strings.Contains(childQueries[0], "*") {
+			for _, v := range childQueries {
+				if err := r.query.AddCounter(v, counter, len(childQueries) > 1); err != nil {
```

Comment:
Should we also remove counters at some moment to avoid leaks?

Reply:
In the line below `query.Close()` I am running the PdhCloseQuery function which should close  all counters contained in the specified query, closes all handles related to the query, and should free all memory associated with the query.
https://docs.microsoft.com/en-us/windows/win32/api/pdh/nf-pdh-pdhclosequery

Reply:
But `query.Close()` is only executed if `AddCounter` fails. Shouldn't we do some cleanup of old counters during the normal operation?

Reply:
I have removed the close query and  added the option to compare counters and remove older ones.

Reply:
Thanks :+1: